File: app/src/main/java/crabster/rudakov/sberschoollesson9hwk/Paint2DView.java
//package crabster.rudakov.sberschoollesson9hwk;
//
//import android.content.Context;
//import android.graphics.Canvas;
//import android.graphics.Color;
//import android.graphics.Paint;
//import android.graphics.Path;
//import android.graphics.PointF;
//import android.util.AttributeSet;
//import android.view.MotionEvent;
//import android.view.View;
//
//import androidx.annotation.Nullable;
//
//import java.util.ArrayList;
//import java.util.List;
//
//import crabster.rudakov.sberschoollesson9hwk.figures.Curve;
//import crabster.rudakov.sberschoollesson9hwk.figures.Figure;
//import crabster.rudakov.sberschoollesson9hwk.figures.Line;
//import crabster.rudakov.sberschoollesson9hwk.figures.Rectangle;
//
//public class Paint2DView extends View {
//
//    private static final float STROKE_WIDTH = 10f;
//
//    private Figure figure;
//
//    private Path path = new Path();
//    private final Paint paint = new Paint();
//    private final List<Figure> figures = new ArrayList<>();
//
//    private boolean isCheckedRectangle = false;
//    private boolean isCheckedLine = false;
//    private boolean isCheckedCurve = false;
//    private boolean isCheckedWhite = false;
//    private boolean isCheckedRed = false;
//    private boolean isCheckedYellow = false;
//
//    public Paint2DView(Context context, @Nullable AttributeSet attrs) {
//        super(context, attrs);
//    }
//
//    /**
//     * Метод отрисовывает фигуры из списка
//     */
//    @Override
//    protected void onDraw(Canvas canvas) {
//        for (Figure x : figures
//        ) {
//            if (x instanceof Rectangle) {
//                float left = Math.min(x.getOrigin().x, x.getCurrent().x);
//                float right = Math.max(x.getOrigin().x, x.getCurrent().x);
//                float top = Math.min(x.getOrigin().y, x.getCurrent().y);
//                float bottom = Math.max(x.getOrigin().y, x.getCurrent().y);
//                int color = x.getColor();
//                paint.setColor(color);
//                canvas.drawRect(left, top, right, bottom, paint);
//            } else if (x instanceof Line) {
//                float startX = x.getOrigin().x;
//                float startY = x.getOrigin().y;
//                float stopX = x.getCurrent().x;
//                float stopY = x.getCurrent().y;
//                int color = x.getColor();
//                paint.setColor(color);
//                canvas.drawLine(startX, startY, stopX, stopY, paint);
//            } else if (x instanceof Curve) {
//                int color = x.getColor();
//                paint.setColor(color);
//                Path newPath = ((Curve) x).getPath();
//                canvas.drawPath(newPath, paint);
//            }
//        }
//    }
//
//    /**
//     * Метод обрабатывает события касания
//     */
//    @Override
//    public boolean onTouchEvent(MotionEvent event) {
//        int action = event.getAction();
//
//        float x1 = event.getX();
//        float y1 = event.getY();
//        PointF current = new PointF(x1, y1);
//
//        switch (action) {
//            case MotionEvent.ACTION_DOWN:
//                figure = createChosenFigure();
//                if (figure != null) {
//                    setUpPaint(figure.getColor());
//
//                    if (figure instanceof Rectangle || figure instanceof Line) {
//                        figure.setOrigin(current);
//                    }
//
//                    if (figure instanceof Curve) {
//                        path = new Path();
//                        path.moveTo(x1, y1);
//                    }
//                    figures.add(figure);
//                }
//                break;
//            case MotionEvent.ACTION_MOVE:
//                if (figure instanceof Curve) {
//                    path.lineTo(x1, y1);
//                    ((Curve) figure).setPath(path);
//                    invalidate();
//                } else if (figure instanceof Rectangle || figure instanceof Line) {
//                    figure.setCurrent(current);
//                    invalidate();
//                }
//                break;
//            case MotionEvent.ACTION_UP:
//            case MotionEvent.ACTION_CANCEL:
//                break;
//        }
//        return true;
//    }
//
//    /**
//     * Метод устанавливает параметры рисования, устанавливая переданный цвет
//     */
//    private void setUpPaint(int color) {
//        paint.setAntiAlias(true);
//        paint.setColor(color);
//        paint.setStrokeWidth(STROKE_WIDTH);
//        paint.setStyle(Paint.Style.STROKE);
//    }
//
//    /**
//     * Метод очищает список фигур и перерисовывает View
//     */
//    public void reset() {
//        figures.clear();
//        invalidate();
//    }
//
//    /**
//     * Метод проверяет значения от CheckBox и создаёт фигуру по
//     * заданным параметрам
//     */
//    private Figure createChosenFigure() {
//        Figure figure = null;
//        if (isCheckedRectangle) {
//            if (isCheckedWhite) figure = new Rectangle(Color.WHITE);
//            if (isCheckedRed) figure = new Rectangle(Color.RED);
//            if (isCheckedYellow) figure = new Rectangle(Color.YELLOW);
//        } else if (isCheckedLine) {
//            if (isCheckedWhite) figure = new Line(Color.WHITE);
//            if (isCheckedRed) figure = new Line(Color.RED);
//            if (isCheckedYellow) figure = new Line(Color.YELLOW);
//        } else if (isCheckedCurve) {
//            if (isCheckedWhite) figure = new Curve(Color.WHITE);
//            if (isCheckedRed) figure = new Curve(Color.RED);
//            if (isCheckedYellow) figure = new Curve(Color.YELLOW);
//        }
//        return figure;
//    }
//
//    /**
//     * Создаём сеттеры для получения значений от CheckBox
//     */
//    public void setCheckedRectangle(boolean checkedRectangle) {
//        isCheckedRectangle = checkedRectangle;
//    }
//
//    public void setCheckedLine(boolean checkedLine) {
//        isCheckedLine = checkedLine;
//    }
//
//    public void setCheckedCurve(boolean checkedCurve) {
//        isCheckedCurve = checkedCurve;
//    }
//
//    public void setCheckedWhite(boolean checkedWhite) {
//        isCheckedWhite = checkedWhite;
//    }
//
//    public void setCheckedRed(boolean checkedRed) {
//        isCheckedRed = checkedRed;
//    }
//
//    public void setCheckedYellow(boolean checkedYellow) {
//        isCheckedYellow = checkedYellow;
//    }
//
//}

File: app/src/main/java/crabster/rudakov/sberschoollesson9hwk/Paint2DView.kt
package crabster.rudakov.sberschoollesson9hwk

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import crabster.rudakov.sberschoollesson9hwk.figures.Curve
import crabster.rudakov.sberschoollesson9hwk.figures.Figure
import crabster.rudakov.sberschoollesson9hwk.figures.Line
import crabster.rudakov.sberschoollesson9hwk.figures.Rectangle
import java.util.ArrayList
import kotlin.math.max
import kotlin.math.min

class Paint2DView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {

    private var figure: Figure? = null
    private var path = Path()
    private val paint = Paint()
    private val figures: MutableList<Figure> = ArrayList()
    private var isCheckedRectangle = false
    private var isCheckedLine = false
    private var isCheckedCurve = false
    private var isCheckedWhite = false
    private var isCheckedRed = false
    private var isCheckedYellow = false

    /**
     * Метод отрисовывает фигуры из списка
     */
    override fun onDraw(canvas: Canvas) {
        for (x in figures) {
            when (x) {
                is Rectangle -> {
                    val left = min(x.getOrigin().x, x.getCurrent().x)
                    val right = max(x.getOrigin().x, x.getCurrent().x)
                    val top = min(x.getOrigin().y, x.getCurrent().y)
                    val bottom = max(x.getOrigin().y, x.getCurrent().y)
                    val color = x.getColor()
                    paint.color = color
                    canvas.drawRect(left, top, right, bottom, paint)
                }
                is Line -> {
                    val startX = x.getOrigin().x
                    val startY = x.getOrigin().y
                    val stopX = x.getCurrent().x
                    val stopY = x.getCurrent().y
                    val color = x.getColor()
                    paint.color = color
                    canvas.drawLine(startX, startY, stopX, stopY, paint)
                }
                is Curve -> {
                    val color = x.getColor()
                    paint.color = color
                    val newPath = x.getPath()
                    canvas.drawPath(newPath, paint)
                }
            }
        }
    }

    /**
     * Метод обрабатывает события касания
     */
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val action = event.action
        val x1 = event.x
        val y1 = event.y
        val current = PointF(x1, y1)
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                figure = createChosenFigure()
                if (figure != null) {
                    setUpPaint(figure!!.getColor())
                    if (figure is Rectangle || figure is Line) {
                        figure!!.setOrigin(current)
                    }
                    if (figure is Curve) {
                        path = Path()
                        path.moveTo(x1, y1)
                    }
                    figures.add(figure!!)
                }
            }
            MotionEvent.ACTION_MOVE -> if (figure is Curve) {
                path.lineTo(x1, y1)
                (figure as Curve).setPath(path)
                invalidate()
            } else if (figure is Rectangle || figure is Line) {
                figure!!.setCurrent(current)
                invalidate()
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
            }
        }
        return true
    }

    /**
     * Метод устанавливает параметры рисования, устанавливая переданный цвет
     */
    private fun setUpPaint(color: Int) {
        paint.isAntiAlias = true
        paint.color = color
        paint.strokeWidth = STROKE_WIDTH
        paint.style = Paint.Style.STROKE
    }

    /**
     * Метод очищает список фигур и перерисовывает View
     */
    fun reset() {
        figures.clear()
        invalidate()
    }

    /**
     * Метод проверяет значения от CheckBox и создаёт фигуру по
     * заданным параметрам
     */
    private fun createChosenFigure(): Figure? {
        var figure: Figure? = null
        if (isCheckedRectangle) {
            if (isCheckedWhite) figure = Rectangle(Color.WHITE)
            if (isCheckedRed) figure = Rectangle(Color.RED)
            if (isCheckedYellow) figure = Rectangle(Color.YELLOW)
        } else if (isCheckedLine) {
            if (isCheckedWhite) figure = Line(Color.WHITE)
            if (isCheckedRed) figure = Line(Color.RED)
            if (isCheckedYellow) figure = Line(Color.YELLOW)
        } else if (isCheckedCurve) {
            if (isCheckedWhite) figure = Curve(Color.WHITE)
            if (isCheckedRed) figure = Curve(Color.RED)
            if (isCheckedYellow) figure = Curve(Color.YELLOW)
        }
        return figure
    }

    /**
     * Создаём сеттеры для получения значений от CheckBox
     */
    fun setCheckedRectangle(checkedRectangle: Boolean) {
        isCheckedRectangle = checkedRectangle
    }

    fun setCheckedLine(checkedLine: Boolean) {
        isCheckedLine = checkedLine
    }

    fun setCheckedCurve(checkedCurve: Boolean) {
        isCheckedCurve = checkedCurve
    }

    fun setCheckedWhite(checkedWhite: Boolean) {
        isCheckedWhite = checkedWhite
    }

    fun setCheckedRed(checkedRed: Boolean) {
        isCheckedRed = checkedRed
    }

    fun setCheckedYellow(checkedYellow: Boolean) {
        isCheckedYellow = checkedYellow
    }

    companion object {
        private const val STROKE_WIDTH = 10f
    }

}